src/rgw/rgw_rados.cc

```diff
@@ -8081,12 +8081,19 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
   }
 
   const uint32_t shard_count = oids.size();
-  const uint32_t num_entries_per_shard =
-    attempt > 0 ?
-    std::min(num_entries,
-	     (attempt *
-	      calc_ordered_bucket_list_per_shard(num_entries, shard_count))) :
-    calc_ordered_bucket_list_per_shard(num_entries, shard_count);
+  uint32_t num_entries_per_shard;
+  if (attempt == 0) {
+    num_entries_per_shard =
+      calc_ordered_bucket_list_per_shard(num_entries, shard_count);
+  } else if (attempt <= 11) {
+    // we'll max out the exponential multiplication factor at 1024 (2<<10)
```

Comment:
is there a discrete gap size which defeats this logic?

Reply:
There shouldn't be. It's worked for over 3000 incomplete multipart uploads (killed the multipart upload after it got to part #490; and did that 7 times before listing directory) before it can get to the actual items of interest.